xwiki-rendering-macros/xwiki-rendering-macro-toc/src/main/java/org/xwiki/rendering/internal/macro/toc/TocMacro.java

```diff
@@ -123,7 +130,15 @@ public boolean supportsInlineMode()
 
         TreeParametersBuilder builder = new TreeParametersBuilder();
         TreeParameters treeParameters = builder.build(rootBlock, parameters, context);
-        return this.tocTreeBuilder.build(treeParameters);
+        List<Block> parametersList = this.tocTreeBuilder.build(treeParameters);
+        if (parameters.getBox()) {
+            BoxMacroParameters boxParameters = new BoxMacroParameters();
+            boxParameters.setCssClass(context.getId() + "message");
+            parametersList = boxMacro.execute(boxParameters, content, context);
```

Comment:
No you shouldn't execute the macro. Instead you should generate a `MacroBlock` and return it in the result.

Reply:
Note that this allows to declare the dependency in the pom.xml on the Box Macro as runtime only above.

Reply:
How would I generate a MacroBlock while using the box macro?

Reply:
`new MacroBlock("box", ...)`. See the javadoc for MacroBlock to see the parameters you need to pass.

Reply:
I'm really not a fan of the MacroBlock idea since it means you need to serialize a toc macro in wiki syntax (without the box parameter this time) and put it as macro block content.

Reply:
hmm indeed hadn't thought of the fact that MacroBlock only supports String is an issue... Maybe an idea would be to make MacroBlock support List<Block>. However this would also mean adding a new Macro#execute() signature. Could be interesting for the future.

But I agree it's not the right approach for now. I'd put a put comment in the code explaining why we cannot use a MacroBlock for the moment.

Reply:
If I shouldn't use the MacroBlock, what should I explore?

Reply:
I was going to say calling execute() but the problem is the same actually since it expects a String... :) @tmortagne what did you have in mind?

If the only solution is to not reuse the Box Macro then I think I prefer serialization rather than code duplication.

Reply:
You could put the actual "boxing" code in some place where toc macro can reuse it (like a simple static public method in BoxMacro for example or a cleaner component used by both box and code macro).

Reply:
Yes but this requires refactoring and I don't really like this way of reusing. It would be much nicer to see the box macro as a black box and not have too strong dependencies between macros. I really prefer a runtime dep.

Reply:
If we really don't want the serialization then I prefer to close this issue as won't fix since users can still manually use the box macro around the toc macro.

Reply:
It's not really a big refactoring. Having to serialize the content is a big -1 from me, macros should only manipulate blocks.

Reply:
ok and I am -1 to introduce a strong build time dependency between the TOC macro and the Box Macro just for an optional parameter especially since the user should be able to write:

```
{{box...}}
{toc/}}
{{/box}}
```

So I guess this makes the issue a won't fix FTM?

Reply:
If toc and box macro cannot share code the toc macro could have its own stuff. We are only talking about a GroupBlock with a specific class here so it's not a big copy/paste...

Reply:
Do we plan to fix this issue?

Reply:
IMO should should simply implement a box around the tox content directly (without reusing BoxMacro code), it's not like it was a lot of code. When I told you to look at BoxMacro code it was to see how to create a box using a block.

Reply:
I still need to think about it more but I don't like it too much ATM and I'm no longer sure we need this jira issue. I'm leaning towards recommending to use the box macro around the toc macro when needing a box. Same as using the cache macro around other macros when you need a cache.

Reply:
looks like it uses a boxblockbuilder class, should I look into that?

Reply:
Yes that's where the code seems to be located but BoxBlockBuilder handle a lot more use cases that you (toc is never inline, no image reference, no title, etc.) o you need only a fraction of what BoxBlockBuilder (only the GroupBlock with the right css classes in it basically).